arkouda/segarray.py

```diff
@@ -803,5 +803,213 @@ def load(cls, prefix_path, dataset='segarray', segment_suffix='_segments', value
         values = load(prefix_path, dataset=dataset + value_suffix)
         return cls(segments, values)
 
+    def intersect_ma(self, other):
+        """
+        Computes the intersection of 2 SegArrays.
+
+        Parameters
+        ----------
+        other : SegArray
+            SegArray to compute against
+
+        Returns
+        -------
+        SegArray
+            Segments are the 1d intersections of the segments of self and other
+
+        See Also
+        --------
+        pdarraysetops.intersect1d
+
+        Examples
+        --------
+        >>> a = [1, 2, 3, 1, 4]
+        >>> b = [3, 1, 4, 5]
+        >>> c = [1, 3, 3, 5]
+        >>> d = [2, 2, 4]
+        >>> seg_a = ak.SegArray(ak.array([0, len(a)]), ak.array(a+b))
+        >>> seg_b = ak.SegArray(ak.array([0, len(c)]), ak.array(c+d))
+        >>> seg_a.intersect_ma(seg_b)
+        SegArray([
+        [1, 3],
+        [4]
+        ])
+        """
+        from arkouda.pdarraysetops import intersect1d
+
+        a_seg_inds = self.grouping.broadcast(arange(self.size)[self._non_empty])
+        b_seg_inds = other.grouping.broadcast(arange(other.size)[other._non_empty])
+        (new_seg_inds, new_values) = intersect1d([a_seg_inds, self.values], [b_seg_inds, other.values])
+        g = GroupBy(new_seg_inds)
+        # This method does not return any empty resulting segments. We need to add these if they are missing.
+        if g.segments.size == self.size:
+            return SegArray(g.segments, new_values[g.permutation])
+        else:
+            segments = zeros(self.size, dtype=akint64)
+            k, ct = g.count()
+            segments[k] = g.segments
+            truth = ~self._non_empty | ~other._non_empty
+            if truth[-1]:
+                segments[-1] = new_values[g.permutation].size
```

Comment:
Since you only need the size, you could eliminate the gather by doing `segments[-1] = g.permutation.size`

Reply:
Updated to eliminate the gather.